### ISSUE TITLE
HOS related but non-issue specific discussion

### ISSUE THREAD

**Issue by TheHappyKoala** [TheHappyKoala/Harmony-of-the-Spheres#65]

**Comment by TheHappyKoala** [comment]
@HugoGranstrom So this will be where we can discuss things that are not directly related to any issue per se. Will close it so that it won't pollute the issue board.

**Comment by TheHappyKoala** [comment]
Cleaned up the issues board and later on I'm going to grab my chain saw and cut off some branches, too. Are there any branches that you would like to keep? Most of them seem like dead wood to me, but just in case, I figured it would be prudent of me to ask :rofl:

**Comment by HugoGranstrom** [comment]
I've been out in the wilderness of Jämtland the last few days fishing, without a internet connection 😱🤣 Seems like you've been up to a lot these few days. Feels like 100 years have passed by 😅 

The 20x speedup seems really promising 😉👌 Have yet to read up on the rest 😅

**Comment by TheHappyKoala** [comment]
<img width="893" alt="Screen Shot 2019-08-07 at 12 26 19" src="https://user-images.githubusercontent.com/23419647/62616092-01f51e00-b90f-11e9-98a3-11863f595f53.png">
<img width="790" alt="Screen Shot 2019-08-07 at 12 28 01" src="https://user-images.githubusercontent.com/23419647/62616125-146f5780-b90f-11e9-91c0-17a98848abff.png">

How I envy you 😄 : few things I enjoy more than a good hike! Nah, haven't really been up to that much, besides tidying up around the repo. As for the other comments/remarks on my part; it's mostly observations along the lines of the sky is blue, so don't think you have to read through all of my bla, bla, bla 🤣 . It's nice to finally have a place for random discussions. 

Toyed around a great deal with your solver and it's working rather well, but I do have a few questions, observations and ideas regarding the solver itself and the way forward. I'm going to dust off the branch I'm currently working on and merge it with master and then I'll inject some bla bla bla into the further Lambert work issue.

**Comment by TheHappyKoala** [comment]
And no, that's not the final gui for the spacecraft stuff... Just a placeholder so that I could test stuff 🤣

**Comment by HugoGranstrom** [comment]
Keep your cleaning coming 😉👌 

It's looking good 😄🤜🤛 Ask as much as you want (in the correct thread 😜🤣). 

I have a feeling we will have to create a Part 2 of this thread soon as not to make the loading and scrolling time too long 🤣

**Comment by HugoGranstrom** [comment]
I've been thinking of ditching the `RKF` integrator for a long time now. It's broken and inferior to both `RKN64` and `RKN12` in every way imaginable. Am I allowed to break your rule for diversity among integrators this one time? 🤣☠️🧟‍♂️

**Comment by HugoGranstrom** [comment]
And I gotta say I love the ASCII rocket you have in one of the webworkers 😍

**Comment by TheHappyKoala** [comment]
Funny that you're bringing up the subject of ditching stuff, because I'm in a spring cleaning mode and I have some things I've been meaning to discuss with you as I'd like to get rid of them. Yes, do ditch the RKF integrator, and I'm sorry I never got to migrating it; totally forgot! Think it was when refactored the integrators to use H3 that I missed it. I was going to say ditch Euler, too, but for historical reasons it should be in there not to mention the fact it's the base class for all the integrators :rofl:.

So, I've been thinking a lot about the camera settings, and I'm very much in the mood to just get rid of the option to set the position of the camera to a celestial body, my main reasons being as follows:

* It only works well for bodies that aren't inclined. If you run the Pluto scenario and set the position to charon and the focus to Pluto, you'll see how the camera starts tumbling and acting weird, which is kind of lame. I don't like having things that are half-good included, and while I could fix this by rotating the vetors such that they are not inclined as far as three.js is concerned, I feel like it would be an unnecessary overhead, especially as...
* The free camera option allows you to zoom in on and follow a celestial body along its trajectory, and it works REALLY well, is more interactive and yea, fun. It does what setting the camera position to a mass does only so much better. 
* Reduce the complexity of the rather wonky camera logic in the scene module. 

With the spacecraft stuff, I'll be adding a cockpit view, which in turn will have the option of hiding the whole gui except for the cockpit controls. And while in the cockpit, you'll be able look in any direction. Then we have the chase option, which I think is pretty cool. 

What do you think??? Do you often find yourself using that camera feature?

**Comment by TheHappyKoala** [comment]
I also want to get rid of the planet atmospheres. They look like crap from a distance and the calculations that go into rendering them cost quite a lot... Feel shadows would be a cooler thing to have, so if you're sending earth on a collision course with Earth, you'll see its shadow cast on Earth

**Comment by TheHappyKoala** [comment]
This one is a bit more controversial, if you will... But... I've been thinking of ditching the sky dome, i.e the milky way background, entirely. First, it makes the celestial bodies pop out, like the Jovian system without the sky dome... All of a sudden the moons are super easy to see. Second, unless you're behind a planet, you won't see stars in space since you're bathing in sun light. What do you think???

**Comment by TheHappyKoala** [comment]
I would much rather spend our gun powder on getting even higher resolution textures for the planets and so on

**Comment by HugoGranstrom** [comment]
Euler is important in its own way, to show how terrible an integrator can be 🤣

I've never really tinkered with the camera settings other than getting a better overview when the default is a camera on a spacecraft and such (I think the old Mars Hohmann Transfer used it). So go ahead and do the changes you feel is necessary to make HOS a better UX 😄 There is no use in having somewhat working functionality. I hadn't actually noticed that planets had atmospheres 😓 but I have noticed the weird light circle around them from a distance, so they won't be sorrowed for 🤣 The shadows though, they sound really lit! 😎 
The Milky Way is cool, but it doesn't really provide much other then looking cool so I'm with you there. 😉 

Have you run any profiling tools on HOS to see where the perfomance hungry monsters live? (if there exists any of those for TS/JS)

**Comment by TheHappyKoala** [comment]
It's official: the Milkyway STAYS. Looked really nice for the scenarios of planets with moons because they're more closeup, but for the solar system and exoplanet ones... Damn, it was all just too empty. 

Oh yes, most browsers, at least Chrome, have excellent profilers if you open the dev tools. For testing individual code snippets, you can always use something like [jsPerf](https://jsperf.com/). TypeScript compiles to JavaScript, so there's no TypeScript profiler that I know of. 

Sweating with migrating my clients db to the production site right now... Cross your fingers for me!

**Comment by TheHappyKoala** [comment]
Oh Hugo, done. God I hate databases. No, I love them, but importing and exporting databases always makes me shit bricks all over the place.

**Comment by HugoGranstrom** [comment]
👍 right, I can imagine it being a bit scary as well, not to have a background 😱 I myself get a bit dizzy when playing with space simulators that don't have backgrounds now when I think about it. I lose my sense of orientation 😅 

See there 😅 right before my nose. Haven't really bothered looking at all the "advanced" stuff in the devtools 🤣

Databases are good to have when they behave and sit still but otherwise I'm with you: they are a mess 🤣

**Comment by TheHappyKoala** [comment]
Man, I'm starting to feel genuinely sorry for Jupiter:

https://www.cnet.com/news/jupiter-just-got-slammed-by-something-so-big-we-saw-it-from-earth/

What's everybody's issue with Jupiter??? Like, leave Jupiter alone! 

Slept awesomely well and now I'm going to embark on a day of Harmony of the Sphering (my made up verb meaning to actively work on developing HOS).

**Comment by HugoGranstrom** [comment]
Aoch! That must have hurt 😱🤣 That's life as Earth's personal vacuum cleaner

**Comment by TheHappyKoala** [comment]
<img width="840" alt="Screen Shot 2019-08-09 at 23 44 09" src="https://user-images.githubusercontent.com/23419647/62810851-ab026b00-baff-11e9-8cf6-1d8687e4cc43.png">

Getting there with the UI. Tomorrow I'm gonna add a "fuel meter" that will let you keep track of your available fuel. 

What I've spent the majority of the day working on, though, is fighting a loosing battle against Chrome. As it turns out, Chrome throttles web workers, which is to say they have a limited budget that they are not allowed to exceed. Safari does this, too. So if you do the trajectory crunching in the web worker, it is very slow, because you don't even get to 60 iterations a second, which makes fiddling around with trajectories prohibitively boring.

I liked the idea of web workers because it would solve the problem in an elegant way, but I'm at the point where, barring setting up an API for crunching this data which I would gladly do if I didn't have to pay for it, I'm gonna have to take my medicine and solve the problem in a less elegant way. 

Apparently, when you're running WebGL, browsers don't throttle things down, which means that you can run 10 - 20 iterations per tick () in the simulation without a hickup. This is what I did first when I told you about the speeding up, just to be able to test things quickly, but little did I know about the throttling. So in the end I'm going to implement what we need this way, even if it pains me. But I'm gonna think hard about how to encapsulate this logic so as not to make the scene module a h00ting mess. 

Will push and merge tomorrow. Won't be the done deal, but I think it'll make your work easier to be able to quickly test things with the UI.

But cheer up Darrell, there's a solution, even if it's not as elegant as you hoped it to be 🤣

**Comment by TheHappyKoala** [comment]
Ah, no, Hugo, I take it all back!!!!!

**Comment by TheHappyKoala** [comment]
When doing these calculations, you only need to know about the primary and the target, so yea... By canning the other bodies we can cut through time as if though it were made of butter!!! Where perturbations are important, I propose adding an option that lets you include perturbers.  What do you think???

**Comment by TheHappyKoala** [comment]
<img width="765" alt="Screen Shot 2019-08-10 at 11 44 17" src="https://user-images.githubusercontent.com/23419647/62820417-49281c80-bb64-11e9-90ca-a03c8d34d5e7.png">

LOL. My all time organic traffic search query.

**Comment by HugoGranstrom** [comment]
I'm not really following 😅 Do you mean that we only need to simulate the path of the primary and the target in our fast-forward simulations?

**Comment by TheHappyKoala** [comment]
Yea, and if that's not enough, you can add perturbers. so when leaving Earth, it would be Sun, Earth and Jupiter

**Comment by TheHappyKoala** [comment]
Think I'm changing the open source gravity simulator to Gravity Superpower Simulator on the splash screen

**Comment by HugoGranstrom** [comment]
If we can choose the pertrubators in an automatic way so that the user dosn't have to. Then yes. If we for example choose Titan as our target, we must have Saturn as one of the pertrubators.

**Comment by TheHappyKoala** [comment]
For spacecraft scenarios we could just add a flag for moons which denotes their primary, no? And if somebody creates a system of their own, the primary, will be the primary you set when you add the mass???

**Comment by TheHappyKoala** [comment]
That or we could loop over all the masses and check their escape velocity relationships... So the Moon's velocity doesn't exceed the escape velocity of Earth, so we have to take Earth into account

**Comment by HugoGranstrom** [comment]
#1 sounds like a good fix 😉

**Comment by TheHappyKoala** [comment]
LOL. Yea, I hate to say it, but not all pull requests are created equal 🤣

**Comment by TheHappyKoala** [comment]
But if you have any objections whatsoever, do raise them. If we do it in the animation loop we can take all the masses into account, so it's still possible, if awkward. My plan is to, when we've rolled this stuff out, email some hosting companies and ask if they would like to sponsor us with free hosting. Might sound like a long shot but they do that for quite a lot of projects, so it's worth a try

**Comment by HugoGranstrom** [comment]
Let's see how this turns out first 😉 If it works, we roll with it, otherwise we know what our plan B is 🅱️ 
That sounds brilliant 😄 I can't imagine this would be much of a burden for them as the majority of the calculations (basically all?) happens on the clientside.

**Comment by TheHappyKoala** [comment]
Aight, deal! And yea, it's just the exoplanet data that's fetched at the moment.

**Comment by HugoGranstrom** [comment]
Phenomenal! 👍 

The UI is looking dank btw 😉

**Comment by HugoGranstrom** [comment]
I have an idea that could solve both determining which Sphere of Influence we are currently in and determining which objects to add as pertrubators. The idea is that we at the beginning construct a tree where the root node is the Sun, it's children are all bodies that are **only** in the Sun's SOI (in other words Planets and lonely comets and asteroids). Then their children are the bodies that are in their SOI (the moons basically). This is recursive so we could potentially have a moon of a moon as well. 
If we use this we don't have to check against all planets and moons to see which SOI the spacecraft is at the moment. If we for example are in the Moon's SOI, we would work our way from the top: 
1. We are in the Sun's SOI.
2. Search all it's children until we find one which we are in. Earth in this case.
3. Search Earth's all children 'til we find one we belong to. If none, we are just in Earth's SOI. In this case we would find the Moon which is a leaf on the tree so we have found our current SOI.

In a similar way we can find all the relevent pertrubating bodies for a spaceflight. A good approach would be to include all bodies which SOIs we are passing through. For example from Earth to Titan: Earth, Sun, Saturn, Titan. If we "look" at the tree we have constructed we see that all of these are bodies we pass if we traverse the tree from Earth up to the Sun, and back down to Titan. 

That may be quite a bit to process so tell me all that you think is unclear 😉

**Comment by TheHappyKoala** [comment]
Makes perfect sense! I'd say let's head down that route. So your idea then is that you set an iteniary where you pick a point of departure and then your final destination in one go and schedule the burns accordingly, taking the whole journey into account, as opposed to dividing the trip into legs where the firs leg is to get within Saturn's sphere of influence, and once there you schedule the burn that will bring you to Titan (I notice you seem to have a thing for Saturn, while I keep on launching my spacecraft to Jupiter :rofl:)?

**Comment by TheHappyKoala** [comment]
As for the algorithm itself; spot on! I :heart:  recursion.

**Comment by HugoGranstrom** [comment]
No we still have to do it like before, this is just a method to calculate which SOI we are in. But now we know how many burns we will do ahead of time.

**Comment by HugoGranstrom** [comment]
Or do you have a clever way of doing it that way? 😄

**Comment by TheHappyKoala** [comment]
No, we're on the same page, and this will come in handy for the gui, having an outline of your trip, especially when you start taking fuel into consideration :oil_drum:

**Comment by TheHappyKoala** [comment]
In the meantime I'm going to try to use the stateToKepler method to try to get some fancy ellipses on the trajectory map that you can see when the camera position is set to cockpit

**Comment by TheHappyKoala** [comment]
Lol, if rotating things in 3D is hard, doing so in 2D with the canvas API is even more so 🤣 . First you need to save the context, then translate the origin of the coordinate system, then apply your rotation and at that point you need to restore the original context because otherwise everything you've drawn will have been rotated.

**Comment by TheHappyKoala** [comment]
I'm alive!!!

**Comment by HugoGranstrom** [comment]
Me too 🤣

**Comment by TheHappyKoala** [comment]
Had to do a code test for a position I've applied for; submitted it today! A lot of devs I know don't particularly like having to do a code test, but I love them, because they always open my eyes to something I would otherwise not even contemplate paying attention to

**Comment by TheHappyKoala** [comment]
Yaaayyyy 🤣

**Comment by TheHappyKoala** [comment]
I've bought node.js hosting with a 20 dollar flat rate, so I'm going to whip up an API for that so that we can do some serious trajectory crunching

**Comment by TheHappyKoala** [comment]
And then I can add the HORIZONS API while I'm at it

**Comment by TheHappyKoala** [comment]
If my virtual server gets "hugged to death", that is a very nice problem to have, I figured, and since the price will never exceed 20 bucks a month, I don't have to worry about some asshat ddosing me into financial hell 🤣

**Comment by TheHappyKoala** [comment]
Gonna wrap up ma branch, first, though... One thing at a time. Quick question: what do you think would be nice to have gui vise for the spaceflight stuff? I'm thinking the relative velocity of your craft and target at rendevouz, fuel and attitude controls (i.e being able to manually set the direction vector of your craft; probably made up that term, so excuse me for insulting the subject you're studying)

**Comment by HugoGranstrom** [comment]
Neat!😄👍 Hope you get it 😉🙏 Is a code test kind of like Fizzbuzz? 🤣 Challenge are made to be conquered 😃 

Awesome! 😁 Haha 🤣 now I'm more motivated then ever to use HOS as much as possible, to squeeze every penny out of it. It wasn't to expensive either. Basically a lunch per month 😋🍔

**Comment by HugoGranstrom** [comment]
That's about what I can come up with as well 😉 "direction vector" is a  completely valid term, otherwise I would have made it into one 🤣

**Comment by TheHappyKoala** [comment]
I've never been asked to type out an algorithm, well, this one time sort of; I was asked to write an algorithm that would automatically resize text so that any number of characters would never exceed a predefined fixed width. It was a funny problem to figure out. 

This and other times I'm usually just asked to write an api with a client, with a filtering table for example, that consumes it. This task was interesting because I was asked to work with Apollo and GraphQL, which I've never done before, so that was really cool

**Comment by TheHappyKoala** [comment]
Ah, phjew, but great, now I can tell people to mind their direction vectors 🤣

**Comment by TheHappyKoala** [comment]
You start school soon, no? What courses will you be taking???

**Comment by HugoGranstrom** [comment]
That sounds really interesting 😄 could see that be useful in for example Photoshop if one wants to fit a piece of text in a certain spot without having to manually test different font sizes. 

Apollo... 🤔🤣 Where have I heard that before? 

I'm kind of starting the 26th so just a week to get this done 😱🤣🤣🤣 I'll be reading about transformation methods (Fourier transforms basically), electronics and a economy course tailored for engineering entrepreneurs I guess 😄 

What does the job you applied for involve?

**Comment by TheHappyKoala** [comment]
Hahaha, I could use that in real life when I'm writing by hand... I always begin typing really large letters when I'm filling out a form, and then I realise I'm running out of space and shrink the letters accordingly, and sometimes I'll even make an arrow pointing to text outside of the field designated for your answer 🤣 

Woho! Wish I could offer you help if you got stuck, but no, I'm officially clueless when it comes to all of the topics you listed... Well, maybe the economics course, but then again I'm a national and not a företagsekonom.... Civilekonom... Always wanted to be a Militärekonom, dock

It's a position at an agency, so it'd be cool if I landed it because I'd get to work on different projects, which is always nice. They work with React and Gatsby, so stack wise I'm a good fit, although I'm not married to a stack per se... At the end of the day it's all JavaScript 👼

**Comment by HugoGranstrom** [comment]
Haha that's so relatable 🤣 

Hopefully my teachers will suffice 🤣 but if they can't solve it, I'll ask you. 'Cause a wrong answer is better than no answer???😅🤷 I have never understood the difference between all the ekonomer 😅🤣 they count money as far as I know (although probably not their own). 

Sounds fun and exciting 😁👍 I can imagine your REACTion when you land the job 😉✌️

**Comment by TheHappyKoala** [comment]
Counting money is what the revisorer do. My business is dealing with scarcity and allocating resources in an optimal way, usually subject to some constraints that create inefficiencies, but are considered to be socially desirable; think minimum wages. I did my first thesis on emissions forecasting and my intention is to make an app out of it at some point, and let people see for themselves what the reality of climate change is given different assumptions about the growth of variables like wealth, population and emissions efficiency per unit of energy, for instance. But arrhhh, time... Time, Hugo! 

Just added fly controls which solves quite a few issues... They can be used for setting the direction vector of the spacecraft, so attitude control, and they let you look around with the mouse or touch/move on mobile devices. Then you use the q and e keys to roll the spacecraft. Need to code up some controls for the rolling since keys aren't readily available on mobile devices, but that should be done in a jiff. 

Quick question for you: I think we should set the tolerance of the simulation depending on whether you are in the sun's or a planet or a moon's sphere of influence. I know that's what an adaptive time step does for you, but I feel like if you have the tolerance set to what makes for smooth journeys between the planets, things go way to fast when you coast around a planet and its moons, even if the simulation remains fairly accurate (RKN12 is such a beast; my jaw hurts from it dropping to the floor on a daily basis in amazement at it 🤣 ). What do you think?

**Comment by HugoGranstrom** [comment]
Hehe I should stay in my field of expertise obviously 🤣 That does sound a lot like engineering, in that there is a lot of problem solving.

Will check out the controls tomorrow 😉 

I agree with you, we don't get much time around the planets with the current implementation 😉 RKN12 is too eager 🤣 let's see how it works out. If it is too sure of itself and spitting out super low errors around planets as well we will have to lower the maxDt instead 🤣 RKN12 is such a troublemaker!

**Comment by TheHappyKoala** [comment]
Man, I've got my butchers knife and I'm in the scene module with a "Here's Johnny!" expression on my face... 🚨 🚨🚨

🎥 beware!

**Comment by HugoGranstrom** [comment]
🤣🤣🤣 better watch out then!

**Comment by TheHappyKoala** [comment]
Sphere of influence stuff works like a charm... Figured running the method every frame would be a waste of resources, so I'm running it once a second; do you think that's an acceptable refresh rate? Suppose it depends on the time step and which body you're hanging out around

**Comment by HugoGranstrom** [comment]
Glad to hear it works 😁 as long as a revolution takes longer than one second that should be fine 😉 if we couple this with your idea of lowering the tolerance around smaller bodies this should work no problem. We could recreate the tree altogether as well say every 30 second if we would have a scenario where stuff moves around a lot.

**Comment by TheHappyKoala** [comment]
Sounds like a reasonable idea... Here's the implementation, in case you're curious: https://github.com/TheHappyKoala/Harmony-of-the-Spheres/commit/c2982e1881c3acfd32af97facda11e1e47dbfbff

**Comment by HugoGranstrom** [comment]
Hmm.... Nested `setInterval` doesn't seem like such a good idea 😅 Maybe it wasn't such a useful idea after all. Why would something other than our spacecraft switch SOI?

**Comment by TheHappyKoala** [comment]
Haha, don't worry!!! It's a hook that is registered when the CockpitDashboard component mounts, and when you unmount it, it is deregistered. It mounts when you open it by setting the camera position to cockpit, and unmounts when you change the camera position to something else. 

You notice how the invocation of the useEffect method returns a method that clears the timer? That method is executed when the component is mounted, just like the timer is setup only when the component is mounted. 

While the component is mounted, it updates the current sphere of influence of the spacecraft and spacecraft only once a second; it also checks if the scenario passed to the component is the same as the scenario registered in the state, and if it isn't, it creates a new SOI tree. 

Without the timer, the current soi check would be run every frame of the simulation, but with it, it is throttled (like stepping on the brake when you're driving or if you are so minded, the gas pedal). 

If you open the element inspector, you will notice that the gui only renders what it needs to render. The markup for the CockpitDashboard disappears when you close it, and it is not only the markup, but all that is associated with it. The same holds true for all components and components nested within them.

**Comment by TheHappyKoala** [comment]
Setting up an autopilot that does a trajectory correction manoeuvre whenever the spacecraft enters a new sphere of influence would be super easy, by the way.

**Comment by HugoGranstrom** [comment]
Thanks for the thorough explanation 🙏😁👍 you truly is the master between us when it comes to this 😉🙇

Yes it should be a piece of cake really 😄✌️ the last part is (optionally) getting it into orbit around the target. Should we aim for a circular orbit?

**Comment by TheHappyKoala** [comment]
Look at getVMag and getOrbit in this module: https://github.com/TheHappyKoala/Harmony-of-the-Spheres/blob/master/src/js/Physics/utils.js

Right now the orbit is set assuming the body is at its apoapsis, but I can add a param so that you can choose whether it is at its peri or apoapsis (mentioning this as most burns are done at periapsis, because of the Oberth effect, which admittedly we're not taking into account, but whatevah 🤣 ). So with this stuff, we could even add orbital elements to the gui so that the user can decide for themselves what kind of orbit they would like around their target. Probably some nonsense going on in that method that is my fault (will look into fixing it), but if you try out the add mass functionality, you can get an idea of what I have in mind 😄. Circular orbit is the special case where the semimajor axis is equal to the distance between the spacecraft and the body it is to orbit. 

Lol, can't say gui development can ever compare to the stuff you're doing. I know just enough physics and modelling to be able the sense the greatness of your contributions. Project wouldn't be where it is without your input and contributions. But that's why I love working with you, our skills are very much complementary.

**Comment by TheHappyKoala** [comment]
That's what I use the ellipse for, btw... So that you can see the orbit of the mass you're adding given the params you've set (in case you haven't stumbled into that tab 🤣 )

**Comment by HugoGranstrom** [comment]
Great 👍😁😉 looks like it will do most of what we want! 

Wise man Darrell 🙇🙏☺️ This really is a team effort and we got one hella of a team! You know more physics than I do gui and react (you did get this whole project rolling on your own for god's sake). I just came and built some nice buildings on your solid Earth 😉🌍 
I'm btw mostly implementing algorithms, piecing together code from various places. Often without fully understanding it 🤣
